TypeScript: automatic typing generation (#1726)
strapi/documentation: docusaurus/docs/dev-docs/configurations/typescript.md

---

> title: TypeScript and autocomplete

Reviewer: I don't think we need to talk about auto-completion in the title

displayed_sidebar: devDocsSidebar
description: How to configure Strapi for TypeScript development. 

---

# TypeScript and autocomplete configuration

[TypeScript](/dev-docs/typescript)-enabled Strapi projects have a specific project structure and handle TypeScript project configuration through [`tsconfig.json` files](#project-structure-and-typescript-specific-configuration-files).

> Strapi also supports types autocompletion in your code editor while you are developing. This feature is related to TypeScript but works for both JavaScript and TypeScript projects. An additional [configuration parameter in the `config/typescript.js|ts` file](/dev-docs/configurations/typescript#strapi-specific-configuration-for-typescript-) allows Strapi to automatically generate types when content-types are updated.

Reviewer: I like the fact that we mention the autogeneration here, but I believe types being exposed from Strapi packages is something that people will be expecting. Even if we're providing way more advanced types (including thanks to dynamic type registries & autogeneration), in the end, TypeScripts libraries are expected to provide auto-completion (including for JS apps). Thus I find it a bit weird that we advertise that as a feature.

| `tsconfig.json` file                      | application root | Manages TypeScript compilation for the server.                                                                                                    |
| `tsconfig.json` file                      | `./src/admin/`   | Manages TypeScript compilation for the admin panel.                                                                                               |

## Strapi-specific configuration for TypeScript

> Strapi supports types autocompletion in your code editor while you are developing. Autocompletion works for both JavaScript and TypeScript projects.

Reviewer: Same as my comment above, maybe we could turn that into smth more like

"while Strapi provides generic typed APIs by default, they can be adapted to reflect the exact state of your application thanks to types auto-generation" (or something like that, you get the idea)

What do you think?

Types generated by Strapi are based on the user project structure. Once the type definitions are emitted into their dedicated files, Strapi reads the type definitions to adapt the autocompletion results accordingly.

To avoid having to [manually generate types](/dev-docs/typescript#generate-typings-for-project-schemas) every time content-types are updated, an optional `config/typescript.js|ts` configuration file can be added, which currently accepts only one parameter:

| Parameter      | Description                                            | Type      | Default |
| -------------- | ------------------------------------------------------ | --------- | ------- |

> | `autogenerate` | Enable or disable automatic types generation as you update content-types | `Boolean` | `false` |

Reviewer: ```suggestion
| `autogenerate` | Enable or disable automatic types generation on server restart | `Boolean` | `false` |
```

| `tsconfig.json` file                      | application root | Manages TypeScript compilation for the server.                                                                                                    |
| `tsconfig.json` file                      | `./src/admin/`   | Manages TypeScript compilation for the admin panel.                                                                                               |

## Strapi-specific configuration for TypeScript <BetaBadge />

> When a Strapi project uses TypeScript, an optional `config/typescript.js|ts` configuration file can be added, which currently accepts only one parameter:

Reviewer: You can use automatic type generation & IntelliSense even without a TS project. Those are two different things

Author: What are types in a non-TypeScript project, then? 🤔

Reviewer: Types generated by Strapi are based on the user application structure and augment Strapi's own APIs (entity-service, etc...).
Thus, we can generate types for a JavaScript application as we're only going to look at the structure. Once the type definitions are emitted into their dedicated files, Strapi will read them to adapt its auto-completion & Intellisense accordingly. This works for JavaScript & TypeScript app.
The configuration file is called `typescript` as it configures **Strapi features** related to TypeScript like autogen. The TypeScript app configuration can be tweaked using the root tsconfig.json files.

| Parameter      | Description                                            | Type      | Default |
| -------------- | ------------------------------------------------------ | --------- | ------- |

> | `autogenerate` | Automatically generate typings as you develop (see [generating typings](/dev-docs/typescript#generate-typings-for-project-schemas)) | `Boolean` | `false` |

Author: ```suggestion
| `autogenerate` | Enable or disable automatic types generation as you update content-types (see [generating typings](/dev-docs/typescript#generate-typings-for-project-schemas)) | `Boolean` | `false` |
```
